### PR TITLE
several changes

### LINKED ISSUE #36: Create HTTPS plugin

### LINKED ISSUE #51: Create redirect method

### PR DESCRIPTION
**Pull request by Tieske**
- cleanup error template
- add redirect method
- add sendFile method (download file) + plugin
- add a TLS plugin

closes #51 
closes #36

### CODE REVIEW

File: src/pegasus/plugins/tls.lua [hunk 1]
```diff
@@ -0,0 +1,55 @@
+--- A plugin that enables TLS (https).
```
Comment (EvandroLG) [inline src/pegasus/plugins/tls.lua:1]: Loved it! ❤️

File: src/pegasus/plugins/tls.lua [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  -- wrap the client socket and replace it
+  client = assert(ssl.wrap(client, params.wrap))
+
+  if params.sni then
+    assert(client:sni(params.sni.names, params.sni.strict))
+  end
+
+  if not client:dohandshake() then
+    print"handshake failed"
+    return false
```
Comment (EvandroLG) [inline src/pegasus/plugins/tls.lua:47]: I'm not sure if it makes sense to return `false` here. Instead, I'd prefer to return `nil`.
WDYT?

File: src/pegasus/response.lua [hunk 3]
```diff
@@ -213,13 +214,58 @@ function Response:write(body, stayOpen)
...
+  return self
+end
+
+-- download by browser, return nil+err if not ok
+function Response:sendFile(path)
+  local filename = path:match("[^/]*$") -- only filename, no path
+  self:addHeader('Content-Disposition', 'attachment; filename="' .. filename .. '"')
+
+  local ok, err = self:writeFile(path, 'application/octet-stream')
+  if not ok then
```
Comment (EvandroLG) [inline src/pegasus/response.lua:257]:
```suggestion
  if err then
```

File: src/pegasus/response.lua [hunk 4]
```diff
@@ -213,13 +214,58 @@ function Response:write(body, stayOpen)
...
+    self:addHeader('Content-Disposition', nil)
+    return nil, err
+  end
+
+  return self
+end
 
+function Response:redirect(location, temporary)
```
Comment (EvandroLG) [inline src/pegasus/response.lua:265]: Awesome! 👍

File: src/pegasus/handler.lua [hunk 5]
```diff
@@ -33,6 +29,18 @@ function Handler:pluginsAlterRequestResponseMetatable()
   end
 end
 
+function Handler:pluginsNewConnection(client)
+  for _, plugin in ipairs(self.plugins) do
+    if plugin.newConnection then
+      client = plugin:newConnection(client)
+      if not client then
+        return false
```
Comment (EvandroLG) [inline src/pegasus/handler.lua:37]: It looks more consistent to me if we return `nil` in these cases.
WDYT?

**Comment by EvandroLG** [comment]
It's just an amazing contribution! It'll be the perfect release of the Pegasus v1! 👍

File: src/pegasus/plugins/tls.lua (hunk 2 above)
Comment (Tieske) [reply src/pegasus/plugins/tls.lua:47]: Rest of the code base returns `stop` add indicator for further processing or stopping. Hence I’ve tried to stick to that.

**Comment by Tieske** [comment]
I have a few more changes to make before a 1.0, but it’s shaping up nicely

File: src/pegasus/response.lua (hunk 3 above)
Comment (Tieske) [reply src/pegasus/response.lua:257]: checking on `ok` is idiomatic in Lua. Only if an expected, positive, result value is either `nil` or `false` the check on `err` is used.

File: src/pegasus/plugins/tls.lua (hunk 2 above)
Comment (Tieske) [inline src/pegasus/plugins/tls.lua:47]:
```suggestion
    return nil
```
nvm my previous comment, this handler is supposed to return the `client`

**Comment by EvandroLG** [comment]
It looks great @Tieske!
Would you like to add something else to this PR? Otherwise, it seems to me the perfect release of the Pegasus v1!

**Comment by Tieske** [comment]
preferably I'd like #131 as well in the 1.0 version.

But I'd like your opinion on the following; should the router be a plugin, or should the router replace the handler? Personally I think the latter, since each router-path should probably get its own plugins.

**Comment by Tieske** [comment]
Lapis is an example.

with a router you typically want specific plugins on specific paths. Eg.

- `/downloads/...` should get a plugin that serves files as downloads
- `/static/...` should get a plugin that serves files as html/text
- `/api/...` should have plugins for content negotiation (if I request `/api/user/evandrolg` I might want to serve JSON if requested by an app, or an html-div if by a browser)
- all of the above should probably have the "compression" plugin

Now how would this work if the router itself is a plugin?

If the router is implemented in the handler, then it could simply default to a route "/" with the "files" plugin first, and a callback as fallback. That would retain the current handler functionality.

**Comment by EvandroLG** [comment]
@Tieske IMO Lapis is not the best example, as they are not an HTTP server but a web framework.
Of course, a web framework like Lapis, Rails or Django should implement its router solution, but I don't see good reasons to have it as a native feature of an HTTP server like Pegasus. IMO, The responsibility of an HTTP server is to implement the server part of the HTTP and HTTPS protocols.

**Comment by Tieske** [comment]
> The responsibility of an HTTP server is to implement the server part of the HTTP and HTTPS protocols.

Good point.

So that means that a router should be either an additional `handler` implementation (instead of a replacement) within the pegasus project, or a separate project altogether (an external "handler" override).

Would the former be acceptable; a second "handler" implementing a router? if not, how did you envision #50  to be implemented?

**Comment by EvandroLG** [comment]
@Tieske I think we could make it available as part of the plugin structure of Pegasus. WDYT? Would you happen to have any other suggestions for it?

**Comment by Tieske** [comment]
can you elaborate on what you mean by "we could make it available as part of the plugin structure of Pegasus"?

**Review by EvandroLG — APPROVED** [review]
LGTM :)

**Comment by EvandroLG** [comment]
@Tieske, let's move the discussion about the Router to the proper PR.
Thanks again! :)

**Pull request merged**